### PR TITLE
Julia Kozeeva Tasks 41-50

### PR DESCRIPTION
**Pull request by usersadvocate**

### CODE REVIEW

File: tasks/42_first-n-smallest/first-n-smallest.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+function firstNSmallest(array, n) {
+  const result = [];
+  const sortedArr = array.slice().sort(function(a, b) {
```
Comment (vovcyan) [inline tasks/42_first-n-smallest/first-n-smallest.js]: Please use arrow functions

**Review by vovcyan — CHANGES_REQUESTED** [review]

File: tasks/42_first-n-smallest/first-n-smallest.js (hunk 1 above)
Comment (usersadvocate) [reply tasks/42_first-n-smallest/first-n-smallest.js]: I fixed it

File: tasks/47_string-expansion/string-expansion.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+function stringExpansion(str) {
+  const regLetter = new RegExp(/^[A-Za-z]+$/);
+  const regNumber = new RegExp(/^[0-9]$/);
+  if (regLetter.test(str)) {
+    return str;
+  }
+  let newStr = '';
+  for (let i = 0; i < str.length; i++) {
+    if (regNumber.test(str[i])) {
+      let j = 1;
```
Comment (vovcyan) [inline tasks/47_string-expansion/string-expansion.js]: What is J ?

Could you please add JSDoc for this functions

**Review by vovcyan — CHANGES_REQUESTED** [review]

File: tasks/47_string-expansion/string-expansion.js (hunk 2 above)
Comment (usersadvocate) [reply tasks/47_string-expansion/string-expansion.js]: I gave the variable a more understandable name

File: tasks/45_equal-sides/equal-sides.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+function findEqualIndex(array) {
+  if (array.slice(1).reduce((a, b) => a + b, 0) === 0) {
+    return 0;
+  }
+  for (let i = 1; i < array.length; i++) {
+    if (array.slice(0, i).reduce((a, b) => a + b, 0) ===
```
Comment (savchukm) [inline tasks/45_equal-sides/equal-sides.js:3]: I think you can include special case written above into this cycle

File: tasks/46_running-average/running-average.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+function runningAverage() {
+  const saveInput = [];
+  return function(num) {
+    saveInput.push(num);
+    return saveInput.reduce((a, b) => a + b)/saveInput.length;
```
Comment (savchukm) [inline tasks/46_running-average/running-average.js]: Spaces around binary operators

File: tasks/47_string-expansion/string-expansion.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  const regLetter = new RegExp(/^[A-Za-z]+$/);
+  const regNumber = new RegExp(/^[0-9]$/);
+  if (regLetter.test(str)) {
+    return str;
+  }
+  let newStr = '';
+  for (let i = 0; i < str.length; i++) {
+    if (regNumber.test(str[i])) {
+      let charIndex = 1;
+      while ((regLetter.test(str[i+charIndex])) && (i+charIndex)<str.length) {
```
Comment (savchukm) [inline tasks/47_string-expansion/string-expansion.js]: Spaces

File: tasks/48_follow-the-spy/follow-the-spy.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+function followTheSpy(arr) {
+  function arrayRemove(arr, value) {
+    return arr.filter(function(el) {
+      return el != value;
```
Comment (savchukm) [inline tasks/48_follow-the-spy/follow-the-spy.js]: Strict equal is most preferable

Comment (savchukm) [reply tasks/48_follow-the-spy/follow-the-spy.js]: Also please use arrow functions

File: tasks/48_follow-the-spy/follow-the-spy.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    });
+  }
+  const result = [];
+  result.push(arr[0][0]);
+  let lastPoint = arr[0][1];
+  let lastRoute = arr[0];
+
+  let count = arr.length;
+
+  while (count>0) {
```
Comment (savchukm) [inline tasks/48_follow-the-spy/follow-the-spy.js]: Spaces round binary operators

File: tasks/49_cat-and-dog-years/cat-and-dog-years.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+function humanYearsCatYearsDogYears(humanYears) {
+  let catYears = 15;
+  let dogYears = 15;
+  if (humanYears >= 2) {
+    catYears += 9;
+    dogYears += 9;
+  }
+  if (humanYears > 2) {
```
Comment (savchukm) [inline tasks/49_cat-and-dog-years/cat-and-dog-years.js]: This if block looks similar to the previous one, I think you need to combine it somehow

**Review by savchukm — CHANGES_REQUESTED** [review]

**Review by savchukm — DISMISSED** [review]

File: tasks/41_sort-by-bits/sort-by-bits.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+function sortByBits(arr) {
+  function countOfBits(num) {
```
Comment (alexkond) [inline tasks/41_sort-by-bits/sort-by-bits.js]: Consider bringing this function to the outer scope.

File: tasks/42_first-n-smallest/first-n-smallest.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+function firstNSmallest(array, n) {
+  const result = [];
+  const sortedArr = array.slice().sort((a, b) => a - b);
+  const smallestNumbers = sortedArr.slice(0, n);
+  array.forEach((num) => {
+    if (smallestNumbers.includes(num)) {
```
Comment (alexkond) [inline tasks/42_first-n-smallest/first-n-smallest.js]: Looks like you can use `filter()` here instead.

File: tasks/47_string-expansion/string-expansion.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ * @return  {String} Returns string without numeric characters
+ */
+function stringExpansion(str) {
+  const regLetter = new RegExp(/^[A-Za-z]+$/);
+  const regNumber = new RegExp(/^[0-9]$/);
+  if (regLetter.test(str)) {
+    return str;
+  }
+  let newStr = '';
+  for (let i = 0; i < str.length; i++) {
```
Comment (alexkond) [inline tasks/47_string-expansion/string-expansion.js]: There's no need to iterate over the string. Look closely to `Regexp`. It has matching groups which could help you.

Comment (alexkond) [reply tasks/47_string-expansion/string-expansion.js]: I sound like magic character that guides you through your quest 🤣

File: tasks/48_follow-the-spy/follow-the-spy.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
+function followTheSpy(arr) {
+  function arrayRemove(arr, value) {
```
Comment (alexkond) [inline tasks/48_follow-the-spy/follow-the-spy.js]: Consider bringing this function to the outer scope.

**Review by alexkond — CHANGES_REQUESTED** [review]

File: tasks/42_first-n-smallest/first-n-smallest.js (hunk 10 above)
Comment (usersadvocate) [reply tasks/42_first-n-smallest/first-n-smallest.js]: I even got rid of array initialization this way, thanks)

File: tasks/47_string-expansion/string-expansion.js (hunk 11 above)
Comment (usersadvocate) [reply tasks/47_string-expansion/string-expansion.js]: at the moment I’m thinking about a new solution

Comment (usersadvocate) [reply tasks/47_string-expansion/string-expansion.js]: I tried to fix it

**Review by savchukm — DISMISSED** [review]

File: tasks/44_time-simulation/time-simulation.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+class SimTime {
```
Comment (vovcyan) [inline tasks/44_time-simulation/time-simulation.js]: Please see task description

function SimTime() {
  // ...
}

let sim = new SimTime();

**Review by vovcyan — CHANGES_REQUESTED** [review]

File: tasks/44_time-simulation/time-simulation.js (hunk 13 above)
Comment (usersadvocate) [reply tasks/44_time-simulation/time-simulation.js]: fixed it

File: tasks/44_time-simulation/time-simulation.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
+function SimTime() {
+  let realTime = 0;
+  let simTime = 0;
+  let speed = 1;
+
+  SimTime.prototype.setSpeed = function(newSpeed) {
```
Comment (savchukm) [inline tasks/44_time-simulation/time-simulation.js]: `this.setSpeed = ...` will be enough

Comment (usersadvocate) [reply tasks/44_time-simulation/time-simulation.js]: fixed it

**Review by savchukm — DISMISSED** [review]

**Review by savchukm — APPROVED** [review]

**Review by vovcyan — APPROVED** [review]

**Review by alexkond — APPROVED** [review]

**Pull request merged**